### PR TITLE
Train WildFire

### PR DESCRIPTION
**Pull request by MateoLostanlen**
The purpose of this PR is to add a basic training loop for the wildfire dataset. This is far from being optimal in terms of performance but it can be considered as a first baseline.

### CODE REVIEW

File: pyronear/datasets/wildfire/split_strategy.py [hunk 1]
```diff
@@ -69,12 +69,22 @@ def split(self, dataset, ratios, seed=42):
...
+            for fire_id_test in fire_ids['test']:
+                del self._fire_id_to_size_to_exhaust[fire_id_test]
+
+        else:
+            fire_ids = {'train': self._get_fire_ids_for_one_split(n_samples_train)}
+            fire_ids['val'] = [id_ for id_ in self._fire_id_to_size if id_ not in fire_ids['train']]
+            # Finish exhaustion
+            for fire_id_test in fire_ids['val']:
+                del self._fire_id_to_size_to_exhaust[fire_id_test]
+            fire_ids['test'] = []
```
Comment (frgfm) [inline pyronear/datasets/wildfire/split_strategy.py]: Isn't this supposed to be in #63 ?

File: pyronear/datasets/wildfire/wildfire.py [hunk 2]
```diff
@@ -158,10 +158,10 @@ def fit(self, wildfire):
...
         self.wildfire = wildfire
         # Some checks first
-        if wildfire.metadata['fire_id'].nunique() != wildfire.metadata['fire_id'].max() + 1:
-            warnings.warn(f"Inconsistent Fire Labeling. Maybe try to label the fire again\n"
-                          f"Distinct values of ids({wildfire.metadata['fire_id'].nunique()}"
-                          f" ≠ {wildfire.metadata['fire_id'].max() + 1})", Warning)
+        # if wildfire.metadata['fire_id'].nunique() != wildfire.metadata['fire_id'].max() + 1:
+        #     warnings.warn(f"Inconsistent Fire Labeling. Maybe try to label the fire again\n"
+        #                   f"Distinct values of ids({wildfire.metadata['fire_id'].nunique()}"
+        #                   f" ≠ {wildfire.metadata['fire_id'].max() + 1})", Warning)
```
Comment (frgfm) [inline pyronear/datasets/wildfire/wildfire.py]: Same here

File: test/test_datasets_wildfire_split.py [hunk 3]
```diff
@@ -93,6 +93,15 @@ def test_inconsistent_ratios_raise_exception(self):
...
 
+    def test_splitting_with_test_to_zero(self):
+        ratios = {'train': 0.81, 'val': 0.19, 'test': 0}
+
+        splitter = WildFireSplitter(ratios, seed=42)
+        splitter.fit(self.wildfire)
+
+        for (set_, ratio_) in splitter.ratios_.items():
+            self.assertAlmostEqual(ratio_, ratios[set_], places=2)
+
```
Comment (frgfm) [inline test/test_datasets_wildfire_split.py]: Same here

**Review by frgfm — CHANGES_REQUESTED** [review]
Thanks for the PR!
There seems to be an issues with overlapping modifications with #63. Could you take a look please?

File: pyronear/datasets/wildfire/split_strategy.py (hunk 1 above)
Comment (MateoLostanlen) [reply pyronear/datasets/wildfire/split_strategy.py]: Yes it's in #63, I wrote this PR on top of #63. So we need to merge #63 before this one

File: references/classification/WildFire/torch/train.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+            args.device = 'cuda:0'
+        else:
+            args.device = 'cpu'
+
+    #Create Dataloaders
+
+    normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
+
+    train_transforms = transforms.Compose([
+        transforms.RandomResizedCrop(size=args.resize // 7, scale=(0.8, 1.0)),
```
Comment (frgfm) [inline references/classification/WildFire/torch/train.py]: Sounds like a pretty big loss of information with a relative output size of  `1/7`. Perhaps we should directly resize to target size?

File: references/classification/WildFire/torch/train.py [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+        transforms.RandomRotation(degrees=15),
+        transforms.ColorJitter(),
+        transforms.RandomHorizontalFlip(),
+        transforms.CenterCrop(size=args.resize),  # Image net standards
```
Comment (frgfm) [inline references/classification/WildFire/torch/train.py]: Generally CenterCrop is used for eval but not train set. Having two crop transformations seems unnecessary. It might be better to stick with RandomResizedCrop but discard CenterCrop

File: references/classification/WildFire/torch/train.py [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+        transforms.ToTensor(),
+        normalize
+    ])
+
+    val_transforms = transforms.Compose([
+        transforms.Resize(size=args.resize // 7),
```
Comment (frgfm) [inline references/classification/WildFire/torch/train.py]: same here for the small resize

**Review by frgfm — CHANGES_REQUESTED** [review]
Thanks for the PR! A few changes and some transformation choices that I'm having doubts about, let me know your thoughts about the comments!

File: references/classification/WildFire/torch/train.py (hunk 4 above)
Comment (MateoLostanlen) [reply references/classification/WildFire/torch/train.py]: It was a mistake, I didn't want to lose so much information. Thank you, that's corrected. I also deleted one of the two crops.

File: references/classification/WildFire/torch/train.py [hunk 7]
```diff
@@ -158,18 +158,16 @@ def main(args):
...
         transforms.RandomHorizontalFlip(),
-        transforms.CenterCrop(size=args.resize),  # Image net standards
         transforms.ToTensor(),
         normalize
     ])
 
     val_transforms = transforms.Compose([
-        transforms.Resize(size=args.resize // 7),
-        transforms.CenterCrop(size=args.resize),
+        transforms.Resize(size=args.resize),
```
Comment (frgfm) [inline references/classification/WildFire/torch/train.py:170]: Small difference here: `RandomResizedCrop` does crop **then** resize the rop.
But here we need the CenterCrop after the Resize. The reason behind it (https://pytorch.org/docs/stable/torchvision/transforms.html#torchvision.transforms.Resize) : if you pass an int to Resize, it keeps the aspect ratio and resize the smallest side to `args.resize`. If you pass a tuple it will resize the entire image to target shape (potentially changing the aspect ratio).

On eval, you probably don't want to stretch the image, but you need them to be all of the same size to batch them. So keep the `transforms.Resize` as is but add back the CenterCrop as well to make sure the image is not stretched !

**Review by frgfm — COMMENTED** [review]
Small change to make for eval transformations but the rest looks good!

File: references/classification/WildFire/torch/train.py (hunk 7 above)
Comment (MateoLostanlen) [reply references/classification/WildFire/torch/train.py:170]: I should resize using a tuple to keep the ratio and then do a centercrop?

Comment (frgfm) [reply references/classification/WildFire/torch/train.py:170]: I was rather suggesting doing resize with an int (`args.resize`) then doing a `CenterCrop(arg.resize)`.
If the image is squared, CenterCrop won't do a thing, if it isn't, it will make it square by cropping

Comment (MateoLostanlen) [reply references/classification/WildFire/torch/train.py:170]: oh yes sorry, Done !

File: pyronear/datasets/wildfire/split_strategy.py (hunk 1 above)
Comment (frgfm) [reply pyronear/datasets/wildfire/split_strategy.py]: Could you cancel out this modif please?

File: test/test_datasets_wildfire_split.py (hunk 3 above)
Comment (frgfm) [inline test/test_datasets_wildfire_split.py]: Could you cancel out this modification? The rest of the PR looks good!

**Review by frgfm — COMMENTED** [review]
Only some side modifications to cancel out, the rest looks good!

**Review by Akilditu — COMMENTED** [review]
Thanks for the PR! 

Added some comments to let the ability of optimizer change and tensor shape for targets.

File: references/classification/WildFire/torch/train.py [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+        model.load_state_dict(torch.load(args.resume)['model'])
+
+    # Send to device
+    model.to(args.device)
+
+    # Loss function
+    criterion = nn.CrossEntropyLoss()
+
+    # optimizer
+    optimizer = optim.Adam(model.parameters(),
```
Comment (Akilditu) [inline references/classification/WildFire/torch/train.py:213]: I would recommend to let the ability to use other Optimizers than Adam

File: references/classification/WildFire/torch/train.py [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+    os.environ['PYTHONHASHSEED'] = str(seed)
+    np.random.seed(seed)
+    torch.manual_seed(seed)
+    torch.cuda.manual_seed(seed)
+    torch.cuda.manual_seed_all(seed)  # if you are using multi-GPU.
+    torch.backends.cudnn.benchmark = False
+    torch.backends.cudnn.deterministic = True
+
+
+def train_batch(model, x, target, optimizer, criterion):
```
Comment (Akilditu) [inline references/classification/WildFire/torch/train.py:44]: Does target needs to be squeeze() here ?

Comment (frgfm) [reply references/classification/WildFire/torch/train.py:44]: Not sure why it would? The model will need the batch dimension, so we only `squeeze` it if it was already `unsqueeze`. If I'm not mistaking, it's not needed then

File: references/classification/WildFire/torch/train.py (hunk 8 above)
Comment (frgfm) [reply references/classification/WildFire/torch/train.py:213]: Agreed, but perhaps we can introduce this in another PR?

Comment (MateoLostanlen) [reply references/classification/WildFire/torch/train.py:213]: Yes let's do that

File: references/classification/WildFire/torch/train.py (hunk 9 above)
Comment (MateoLostanlen) [reply references/classification/WildFire/torch/train.py:44]: It depend of your dataloader, but in this script there is no need to squeeze the target

File: pyronear/datasets/wildfire/split_strategy.py (hunk 1 above)
Comment (MateoLostanlen) [reply pyronear/datasets/wildfire/split_strategy.py]: Done !

**Review by frgfm — APPROVED** [review]
Thanks a lot for the changes!
